packages/admin-ui/src/calendar/calendar-state.ts

```diff
@@ -0,0 +1,320 @@
+import type { Dispatch, SetStateAction } from 'react'
+import { useMemo, useState, useCallback } from 'react'
+import { unstable_useId as useId } from 'reakit'
+import {
+  addDays,
+  addMonths,
+  addWeeks,
+  addYears,
+  endOfMonth,
+  getDaysInMonth,
+  isSameMonth,
+  startOfDay,
+  startOfMonth,
+  subDays,
+  subMonths,
+  subWeeks,
+  subYears,
+} from 'date-fns'
+
+import { useWeekStart } from '../i18n'
+import { useWeekDays, generateDaysInMonthArray, toUTCString } from './utils'
+import type { InputState } from '../types'
+
+export function useCalendarState(
+  props: CalendarInitialState = {}
+): CalendarStateReturn {
+  const {
+    value: initialValue = toUTCString(new Date()),
+    minValue,
+    maxValue,
+    isDisabled = false,
+    isReadOnly = false,
+    autoFocus = false,
+  } = props
+
+  const [value, setValue] = useState(initialValue)
+
+  const date = useMemo(() => new Date(value), [value])
+  const minDateValue = useMemo(
+    () => (minValue ? new Date(minValue) : new Date(-864e13)),
+    [minValue]
+  )
+
+  const maxDateValue = useMemo(
+    () => (maxValue ? new Date(maxValue) : new Date(864e13)),
+    [maxValue]
+  )
+
+  const [currentMonth, setCurrentMonth] = useState(date)
+  const [focusedDate, setFocusedDate] = useState(date)
+  const [isFocused, setFocused] = useState(autoFocus)
+  const month = currentMonth.getMonth()
+  const year = currentMonth.getFullYear()
+  const weekStart = useWeekStart()
+  const weekDays = useWeekDays(weekStart)
+
+  // Get 2D Date arrays in 7 days a week format
+  const daysInMonth = useMemo(() => {
+    let monthStartsAt = (startOfMonth(currentMonth).getDay() - weekStart) % 7
+
+    if (monthStartsAt < 0) {
+      monthStartsAt += 7
+    }
+
+    const days = getDaysInMonth(currentMonth)
+    const weeksInMonth = Math.ceil((monthStartsAt + days) / 7)
+
+    return generateDaysInMonthArray(month, monthStartsAt, weeksInMonth, year)
+  }, [month, year, currentMonth, weekStart])
+
+  const isInvalidDateRange = useCallback(
+    (value: Date) => {
+      const min = new Date(minDateValue)
```

Comment:
I think `minDateValue` and `maxDateValue` is already a Date, so it wouldn't be necessary to create a` new Date()`.

I also realized that `minDateValue` and `maxDateValue` are used only here, so maybe they could be declared inside the function `isInvalidDateRange` and the dependencies would be only `minValue` and `maxValue`

Reply:
That's a really good point. Suggestion accepted